### PR TITLE
:tada: Destination postgres: Add SSL certificates and update normalization

### LINKED ISSUE #4034: Support Postgres SSL connection
Currently the Postgres connector has an SSL toggle, however there is no where to specify the appropriate information to establish the SSL connection:
- CA certificate.
- Client certificate.
- Client key.
- Client key password.
- SSL Mode.

This is a must for us a all the connections to our Postgres databases require SSL.

### PR DESCRIPTION
**Pull request by andriikorotkov**
## What
Add SSL certificates to destination postgres and destination postgres strict encrypt.

## Recommended reading order
1. `PostgresDestination.java`
2. `spec.json`
3. `expected_spec.json`
4. `PostgresDestinationStrictEncrypt.java`
5. `PostgresDestinationSSLFullCertificateAcceptanceTest.java`
6. `PostgresDestinationStrictEncryptAcceptanceTest.java`
7. `PostgresDestinationTest.java`
8. `transform.py`

## 🚨 User Impact 🚨
Added a dropdown for choosing SSL mode. Added fields for SSL modes verify-full and verify-ca to add root key and client keys, and added a password for custom keystore.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java
Comment (etsybaev) [inline airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java:60]: Why don't we use container creds anymore?

**Review by etsybaev — APPROVED** [review]
Please do not forget to run checkstyle job and pass PR's checklist

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java:60]: because I have configured an SSL connection for the postgres user that I create when starting the container. The Client certificate was created specifically for this user, which I use for tests.

**Review by edgao — COMMENTED** [review]
you mentioned over slack that `testCustomDbtTransformations` is disabled - where is that configured?

I think ideally we'd try to track that down, since it _should_ be supported - if not in this PR, then at least it should be tracked in a followup issue.

Also, were you able to test custom transformations on a local OSS instance? That would give us some confidence that something is broken in the test setup rather than the actual functionality.

**Comment by edgao** [comment]
(the code changes themselves look pretty reasonable, I just want to doublecheck on testing + figure out a plan for getting test coverage)

**Comment by andriikorotkov** [comment]
@edgao, I created a ticket to fix the `TestcustomdbttransFormations` of the test. I add a link to it - https://github.com/airbytehq/airbyte/issues/14916. To this test I have added annotation @disabled in the `PostgresdestinationsslfullCerticateptancetest` Classes - **Line 122**. All other tests work well, you can make sure that when you will look at this pull request.

**Review by grishick — COMMENTED** [review]
This PR needs:

- Description of changes
- Guide on how to read the PR (where to start, etc)
- Rebase to pass CI tests

File: airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/PostgresSourceTest.java
Comment (grishick) [inline airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/PostgresSourceTest.java:484]: this change looks unnecessary

File: airbyte-integrations/connectors/destination-postgres-strict-encrypt/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestinationStrictEncrypt.java
Comment (grishick) [inline airbyte-integrations/connectors/destination-postgres-strict-encrypt/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestinationStrictEncrypt.java:29]: we should also remove "disable" and "allow" modes from the strict encrypt version

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java
Comment (grishick) [inline airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java:36]: why this change is needed?

File: airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/PostgresSourceTest.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/PostgresSourceTest.java:484]: these changes were added automatically after running the code formatting command. I left them so that later developers who will do automatic formatting of the code do not encounter these changes.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java:36]: as I understood from the documentation, it is more correct that the common name (CN) should be different for root certificates and server certificates. For cases like ours, it is recommended to use the IP address for the root certificate, and the localhost for the server certificate. These changes will not break the source or destination tests.

File: airbyte-integrations/connectors/destination-postgres-strict-encrypt/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestinationStrictEncrypt.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-postgres-strict-encrypt/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestinationStrictEncrypt.java]: as I understood from the documentation, it is more correct that the `common name (CN)` should be different for root certificates and server certificates. For cases like ours, it is recommended to use the IP address for the root certificate, and the localhost for the server certificate. These changes will not break the source or destination tests.

**Comment by andriikorotkov** [comment]
> This PR needs:
> 
> * Description of changes
> * Guide on how to read the PR (where to start, etc)
> * Rebase to pass CI tests

@grishick, thank you for these comments. I've updated the description of this pull request, added a list of files with the main changes, and fixed all the merge conflicts that occurred between my branch and the master branch. Now, I will run CI tests for both destinations, I think all tests will run fine.

**Review by edgao — APPROVED** [review]
LGTM, maybe worth noting in the changelog that custom transformations may not fully support SSL certs yet.

---

do we have some way to manually test custom transformations? I was doing a little digging and suspect that there's an actual problem there. Specifically, `airbyte/airbyte-workers/src/main/java/io/airbyte/workers/general/DbtTransformationRunner.java` says something about using the standard NormalizationRunner to configure the dbt project - maybe that somehow is skipping the `.crt` file copy.

(I don't think it blocks this PR though - the previous situation was that nobody could use SSL certs with destination-postgres, and the new situation is that only people using custom transforms can't use SSL certs, which seems like a strict improvement)

**Review by grishick — APPROVED** [review]
please rebase and resolve conflicts before publishing and merging

**Comment by andriikorotkov** [comment]
@edgao I managed to fix the connector to work correctly with customTransformation. Now this test works fine. But my tests fail after changes that affected the formatting of datetime fields. Once that's fixed, I'll be able to publish and merge with these changes.

**Comment by edgao** [comment]
sounds good. The datetime problem seems to be happening on master as well, i.e. is unrelated to this PR - I think @Phlair is trying to solve it

**Comment by andriikorotkov** [comment]
@edgao Great, thank you for the good news. I will wait until this issue is resolved and then I will merge this pull request.

**Comment by edgao** [comment]
@andriikorotkov fyi we have https://github.com/airbytehq/airbyte/issues/15172 to track the destination-postgres test failures. I didn't see an obvious root cause unfortunately.

(cc @grishick )

**Comment by Phlair** [comment]
@andriikorotkov fyi https://github.com/airbytehq/airbyte/pull/14790 is now merged 👍

**Comment by andriikorotkov** [comment]
/publish connector=connectors/destination-postgres run-tests=false

**Comment by andriikorotkov** [comment]
/publish connector=connectors/destination-postgres-strict-encrypt run-tests=false auto-bump-version=false

**Pull request merged**
